Add InvisibleSymbols Fixer

ivan1986 opened the pull request:
fixer for #2601

keradus commented on README.rst:
```diff
@@ -295,6 +295,9 @@ Choose from the list of available rules:
    | Code MUST use configured indentation type.
 
+* **invisible_symbols**
+   | Remove ZWSP, NBSP and other unicode symbols.
```
shortcuts need to be resolved

keradus commented:
Hi !
Thanks for raising it that fast !
It's a great start for a fixer !

IMO, we should make it configurable, allowing to fix any of 3 types:
- comments (remove odd whitespaces inside comments)
- whitespaces (remove odd whitespaces from other whitespaces)
- strings (so case you found initial issue)

that way one may want to fix only whitespaces or whitespaces+comments, but not strings (he could desire odd chars in them in edge cases)

also, be aware that if token would be empty after dropping odd spaces - we should remove whole token (could happen only for `T_WHITESPACE` token)

ivan1986 commented:
@keradus IMO does not need to configure it - if we found this symbols in the code it's a copy-paste mistake or we do some real strange.

keradus commented on src/Fixer/Basic/InvisibleSymbolsFixer.php:
```diff
@@ -0,0 +1,85 @@
+    public function fix(\SplFileInfo $file, Tokens $tokens)
+    {
+        foreach ($tokens as $index => $token) {
+            $token->setContent(strtr($token->getContent(), $this->symbolsReplace));
```
no need for translation in all tokens, please check the tokens first (eg, no need for fixing `T_PRIVATE`)

keradus commented on tests/Fixer/Basic/InvisibleSymbolsFixerTest.php:
```diff
@@ -0,0 +1,54 @@
+     */
+    public function testFix($expected, $input, $file)
+    {
+        $this->doTest($expected, $input, $file);
```
is there a need for passing a file ?

keradus commented on src/Fixer/Basic/InvisibleSymbolsFixer.php:
```diff
@@ -0,0 +1,85 @@
+ *
+ * @author Ivan Boprzenkov <ivan.borzenkov@gmail.com>
+ */
+final class InvisibleSymbolsFixer extends AbstractFixer
```
This rule may change content of vars, thus it's risky (I agree that in most cases it would be a proper fix, but sadly it may be desired to have it in string values in some edge cases).
For that, please mark rule as risky

keradus commented on src/Fixer/Basic/InvisibleSymbolsFixer.php:
```diff
@@ -0,0 +1,85 @@
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * Fixer Zero-width space (ZWSP), Non-breaking space (NBSP) and other unicode symbols.
```
this is unicode symbol: `☯`
how does the rule deal with it ?

SpacePossum commented:
Nice fixer idea, love to see this in.
I was thinking about proposed configuration and just thinking out loud here, maybe we could make generic 'NonPrintableCharacterFixer' (name can change of course) ?
We could than merge the 'line_ending' fixer in this as well.

Than we could add configuration of character mapping, for example:
* `\r\n` => `\n`
* `\r` => ''
* `0xe2` => ' '

We allow configuration for 'T_ENCAPSED_AND_WHITESPACE, T_END_HEREDOC, T_OPEN_TAG, T_WHITESPACE, T_COMMENT, T_DOC_COMMENT, T_START_HEREDOC, any thing else' to give control about what should be fixed (making the fixer risky by configuration).

WDYT?

keradus commented on src/Fixer/Basic/NonPrintableCharacterFixer.php:
```diff
@@ -54,21 +54,33 @@ public function fix(\SplFileInfo $file, Tokens $tokens)
+            null,
+            null,
+            null,
+            'Risky change strings data.'
```
This is not a valid sentence.

keradus commented on tests/Fixer/Basic/NonPrintableCharacterFixerTest.php:
```diff
@@ -22,33 +22,26 @@
+    public function testFix($expected, $input = null)
     {
-        $this->doTest($expected, $input, $file);
+        $this->doTest($expected, $input);
```
FYI: even when the content was read from file, it's not obligatory to pass the file to `doTest`.

ivan1986 commented:
@SpacePossum change name, about tokens - a good solution would be to process all text in one strtr, but I do not know how to do it.

julienfalque commented on src/Fixer/Basic/NonPrintableCharacterFixer.php:
```diff
@@ -0,0 +1,97 @@
+            null,
+            null,
+            null,
+            'Risky if invisible unicode contains in data strings.'
```
What about: `Risky when strings contain intended invisible characters.`?

julienfalque commented on src/Fixer/Basic/NonPrintableCharacterFixer.php:
```diff
@@ -0,0 +1,97 @@
+    public function getPriority()
+    {
+        // must run first (at least before Fixers that using Tokens) - for speed reason of whole fixing process
+        return 90;
```
Is this really required? How does it improve process speed?

keradus commented:
> a good solution would be to process all text in one strtr, but I do not know how to do it.

it's not a good solution. it's slow solution that will have to revisit all content even if we don't want to modify most of it.

julienfalque commented on tests/Fixer/Basic/NonPrintableCharacterFixerTest.php:
```diff
@@ -0,0 +1,47 @@
+        $this->doTest($expected, $input);
+    }
+
+    public function provideExamples()
```
`provideFixCases`?

julienfalque commented on tests/Fixer/Basic/NonPrintableCharacterFixerTest.php:
```diff
@@ -0,0 +1,47 @@
+
+    public function provideExamples()
+    {
+        return array(
```
Could you please add more test cases, with invisible characters in comments for example?

julienfalque commented on src/Fixer/Basic/NonPrintableCharacterFixer.php:
```diff
@@ -0,0 +1,97 @@
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * Fixer Zero-width space (ZWSP), Non-breaking space (NBSP) and other invisible unicode symbols.
```
```diff
-Fixer
+Removes
```

Also, what about adding links to some doc about these symbols? Would that be useful?

keradus commented:
@SpacePossum , about your idea - it may indeed be nice to have next `General***` fixer that could be used inside different fixers, like we already do eg with `GeneralPhpdocAnnotationRemoveFixer`.
Then, this one, tabs to spaces, line endings, could be rewritten to use that `General***`.
But let us keep it outside of scope of this PR.

keradus commented on tests/Fixer/Basic/NonPrintableCharacterFixerTest.php:
```diff
@@ -0,0 +1,55 @@
+    {
+        return array(
+            array(
+                '<?php echo "Hello World !";',
```
not needed, it's autogenerated from next case

ivan1986 replied:
fixed

keradus commented on src/Fixer/Basic/NonPrintableCharacterFixer.php:
```diff
@@ -0,0 +1,88 @@
+        $this->symbolsReplace = array(
+            pack('CCC', 0xe2, 0x80, 0x8b) => '',
+            pack('CCC', 0xe2, 0x80, 0x87) => ' ',
+            pack('CCC', 0xe2, 0x80, 0xaf) => ' ',
```
it is not tested

ivan1986 replied:
added

keradus commented on src/Fixer/Basic/NonPrintableCharacterFixer.php:
```diff
@@ -0,0 +1,88 @@
+    {
+        parent::__construct();
+        $this->symbolsReplace = array(
+            pack('CCC', 0xe2, 0x80, 0x8b) => '',
```
would be nice to provide names of that symbols

ivan1986 replied:
hmm, no ideas, add coments

ivan1986 commented:
added remove empty tokens
added isAnyTokenKindsFound
I not sure about type tokens list in isAnyTokenKindsFound, please check it and if it wrong tell me correct list for fix

SpacePossum commented on src/Fixer/Basic/NonPrintableCharacterFixer.php:
```diff
@@ -0,0 +1,102 @@
+    {
+        foreach ($tokens as $index => $token) {
+            $token->setContent(strtr($token->getContent(), $this->symbolsReplace));
+            if ($token->isEmpty()) {
```
the check and remove is not needed, it's done by the token itself already; https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/2.1/src/Tokenizer/Token.php#L477  (the `remove` method does not exists anyway I think?)

ivan1986 replied:
removed empty

add test for repeat symbols - all ok

keradus replied:
good catch here

SpacePossum commented on src/Fixer/Basic/NonPrintableCharacterFixer.php:
```diff
@@ -0,0 +1,102 @@
+            T_STRING,
+            T_STRING_VARNAME,
+            T_INLINE_HTML,
+            T_WHITESPACE,
```
for a candidate check I would return `count($tokens) > 1` as that would result in the same but a bit faster

keradus replied:
it's not the same. 
and `count($tokens) > 1` is already checked **before** calling `isCandidate` ;)

cc @ivan1986

keradus commented on src/Fixer/Basic/NonPrintableCharacterFixer.php:
```diff
@@ -0,0 +1,88 @@
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        foreach ($tokens as $index => $token) {
+            $token->setContent(strtr($token->getContent(), $this->symbolsReplace));
```
still, https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2603#discussion_r106534435
for me, lack of resolving this one comment is last blocker for this PR, all the rest are just minors

ivan1986 replied:
added list tokens and add if

keradus commented on src/Fixer/Basic/NonPrintableCharacterFixer.php:
```diff
@@ -0,0 +1,88 @@
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        foreach ($tokens as $index => $token) {
```
index is not used

ivan1986 replied:
fixed

keradus commented:
please fix deprecations

keradus commented on src/Fixer/Basic/NonPrintableCharacterFixer.php:
```diff
@@ -0,0 +1,102 @@
+        T_CONSTANT_ENCAPSED_STRING,
+        T_END_HEREDOC,
+        T_OPEN_TAG,
+        T_DOC_COMMENT,
```
if I would remove `T_DOC_COMMENT` test for this fixer would still pass.
it is not good. probably you need to provide more test cases

keradus replied:
I see you added test for `T_DOC_COMMENT`. that's nice, but following are still missing:
```
        T_STRING,
        T_STRING_VARNAME,
        T_INLINE_HTML,
        T_WHITESPACE,
        T_ENCAPSED_AND_WHITESPACE,
        T_END_HEREDOC,
        T_OPEN_TAG,
        T_START_HEREDOC,
```

ivan1986 commented:
broken only in php 7.2 because <7.2 in composer

keradus commented:
That's not true. Read the error log.
Failing because of trailing whitespaces in:
> ./tests/Fixer/Basic/NonPrintableCharacterFixerTest.php:62: ./tests/Fixer/Basic/NonPrintableCharacterFixerTest.php:66: function f(string $p) ./tests/Fixer/Basic/NonPrintableCharacterFixerTest.php:71: ./tests/Fixer/Basic/NonPrintableCharacterFixerTest.php:75: function f(string $p)

julienfalque commented:
@ivan1986 Actually Travis build fails because there are trailing whitespaces like [here](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2603/files#diff-ae2304b53b736acede7e556747d7a9f0R72) (quite ironic 😁)

keradus commented on src/Fixer/Basic/NonPrintableCharacterFixer.php:
```diff
@@ -0,0 +1,100 @@
+    private $symbolsReplace;
+
+    private static $tokens = array(
+        T_NUM_STRING,
```
`T_NUM_STRING` and `T_WHITESPACE` are not covered by tests

keradus commented:
Thank you @ivan1986.
